### PR TITLE
Use SQLDelight instead of Room

### PR DESCRIPTION
**Pull request by Skrilltrax**
Deletes all the existing Room database code and replaces it with a simplified SQLDelight-backed implementation ready for Kotlin Multiplatform.

### CODE REVIEW

**Comment by msfjarvis** [comment]
CI failure is a cache problem

`Error loading androidx migration mapping: org/apache/batik/w3c/dom/ElementTraversal`

lemme see if we can force an invalidation

**Comment by msfjarvis** [comment]
officially out of ideas

**Comment by msfjarvis** [comment]
We're running into https://github.com/cashapp/sqldelight/issues/1343

**Comment by msfjarvis** [comment]
Fix from https://github.com/cashapp/sqldelight/issues/2058 apparently works, let's see.

**Comment by msfjarvis** [comment]
It didn't work, and now I'm losing it. Goodnight.

File: app/src/test/java/dev/msfjarvis/lobsters/data/local/SqlDelightQueriesTest.kt [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    posts.forEach { postQueries.insertOrReplacePost(it) }
+
+    // Delete 2nd post
+    postQueries.deleteAllPosts()
+
+    val postsCount = postQueries.selectCount().executeAsOne()
+
+    // Check if size is 2, and only the correct post is deleted
+    assertEquals(0, postsCount)
+  }
```
Comment (msfjarvis) [inline app/src/test/java/dev/msfjarvis/lobsters/data/local/SqlDelightQueriesTest.kt:165]: This entire test confuses me

Comment (Skrilltrax) [reply app/src/test/java/dev/msfjarvis/lobsters/data/local/SqlDelightQueriesTest.kt:165]: Don't read the comments lol. I copied and updated the code but forgot to update the comment

**Comment by msfjarvis** [comment]
The remaining step here is to figure out how to pass a custom serializer to Ktor, since we can generate one automatically using [this](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/serializers.md#deriving-external-serializer-for-another-kotlin-class-experimental).

**Comment by msfjarvis** [comment]
> The remaining step here is to figure out how to pass a custom serializer to Ktor, since we can generate one automatically using [this](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/serializers.md#deriving-external-serializer-for-another-kotlin-class-experimental).

Blocked on https://github.com/Kotlin/kotlinx.serialization/issues/1192

**Comment by msfjarvis** [comment]
Might have to bring back Moshi since the platforms I wish to support (JVM and Android) both work with it.

**Comment by msfjarvis** [comment]
The new code "looks" good but is broken in more ways than one, and is going to require fixing.

**Comment by msfjarvis** [comment]
Tests pass but the entire thing is :sparkles: broken as fuck :sparkles: so it's now all yours

**Comment by Skrilltrax** [comment]
Broken at the viewmodel layer?

**Comment by msfjarvis** [comment]
> Broken at the viewmodel layer?

Apparently, posts don't load.

**Comment by msfjarvis** [comment]
Making the `is_saved` field non-null makes Moshi expect it from the remote API response, which will obviously not have it. We need to make the transformation at runtime.

**Comment by msfjarvis** [comment]
Rebased on develop

File: app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt [hunk 2]
```diff
@@ -2,17 +2,30 @@ package dev.msfjarvis.lobsters.data.remote
...
   private val lobstersApi: LobstersApi,
+  private val lobstersRepository: LobstersRepository,
 ) : PagingSource<Int, LobstersPost>() {
 
   override suspend fun load(params: LoadParams<Int>): LoadResult<Int, LobstersPost> {
     return try {
       val page = params.key ?: 1
-      val posts = lobstersApi.getHottestPosts(page)
+      val posts = lobstersApi.getHottestPosts(page).map { post ->
+        // We can replace these two lines below with a getOrInsertPost repository method.
```
Comment (msfjarvis) [inline app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt:18]: Why don't we then?

File: app/src/main/java/dev/msfjarvis/lobsters/ui/main/MainActivity.kt [hunk 3]
```diff
@@ -58,6 +58,8 @@ fun LobstersApp() {
   val hottestPosts = viewModel.posts.collectAsLazyPagingItems()
   val savedPosts by viewModel.savedPosts.collectAsState()
 
+  viewModel.getSavedPosts()
```
Comment (msfjarvis) [inline app/src/main/java/dev/msfjarvis/lobsters/ui/main/MainActivity.kt:61]: This is a leaky abstraction, we should have this be done by the VM itself.

File: app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt [hunk 4]
```diff
@@ -5,48 +5,51 @@ import androidx.lifecycle.viewModelScope
...
   private val pagingSource: LobstersPagingSource,
-  database: PostsDatabase,
+  private val lobstersRepository: LobstersRepository,
 ) : ViewModel() {
   private val _savedPosts = MutableStateFlow<List<LobstersPost>>(emptyList())
-  private val savedPostsDao = database.savedPostsDao()
-  val savedPosts: StateFlow<List<LobstersPost>> get() = _savedPosts
-  val posts = Pager(PagingConfig(25)) {
+  val savedPosts = _savedPosts.asStateFlow()
+  val posts = Pager(PagingConfig(5)) {
```
Comment (msfjarvis) [inline app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt:23]: Why this change? This is what loads the data from the remote server, and the page size is 25 in the API.

File: database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package dev.msfjarvis.lobsters.model
+
+import com.squareup.moshi.Moshi
+import com.squareup.sqldelight.ColumnAdapter
+
+class SubmitterAdapter : ColumnAdapter<Submitter, String> {
```
Comment (msfjarvis) [inline database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt:6]: I find it weird that we need this for `Submitter` but not `KeybaseSignature`. Can you check why?

**Review by msfjarvis — DISMISSED** [review]

**Comment by msfjarvis** [comment]
I'll fix the tests in the morning.

**Comment by Skrilltrax** [comment]
Don't fix the tests atm,The changes are not finalized yet.

File: app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt (hunk 2 above)
Comment (Skrilltrax) [reply app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt:18]: Because it is still WIP :eyes:

File: app/src/main/java/dev/msfjarvis/lobsters/ui/main/MainActivity.kt (hunk 3 above)
Comment (Skrilltrax) [reply app/src/main/java/dev/msfjarvis/lobsters/ui/main/MainActivity.kt:61]: Yeah, we can have this in viewModel's init block.

File: app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt (hunk 4 above)
Comment (Skrilltrax) [reply app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt:23]: Yeah, this was a test change that I did to try and fetch lesser number of posts for debugging. This should not have been a part of the commit. I'll revert it along with the other changes that are required.

File: app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt (hunk 2 above)
Comment (Skrilltrax) [reply app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt:18]: Also, one problem that I've noticed is that we don't have a single source of truth for the data. Some data is coming from the api and other is coming from the DB.

I think we should make the DB the single source of truth and load all data from there.

File: database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt (hunk 5 above)
Comment (Skrilltrax) [reply database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt:6]: Looks like our SQLDelight model is not using `KeybaseSignature`. I'll fix it.

**Comment by Skrilltrax** [comment]
@msfjarvis We are persisting only the saved posts, right?

**Comment by msfjarvis** [comment]
> @msfjarvis We are persisting only the saved posts, right?

Yep

**Comment by Skrilltrax** [comment]
We still need `kotlin-reflect` and moshi for the actual `LobstersPost` item.

File: database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt (hunk 5 above)
Comment (msfjarvis) [reply database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt:7]: Nah it's not that. `KeybaseSignature` is part of `Submitter`, and its serialization is covered by the JSON adapter generated for `Submitter`.

File: app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt (hunk 2 above)
Comment (msfjarvis) [reply app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt:18]: I'm not too sure about having the main stream being persisted to the database. It's nice as an offline fallback but invalidation is tricky.

**Comment by msfjarvis** [comment]
I tried switching to [moshi-metadata-reflect](https://github.com/ZacSweers/MoshiX/tree/main/moshi-metadata-reflect) [here](https://msfjarvis.dev/g/compose-lobsters/34d7ed81ddb7) but it didn't seem to work (crashes with failure to resolve kotlin-reflect), if it's interesting to you then might be worth fixing and including here.

**Comment by msfjarvis** [comment]
> We still need `kotlin-reflect` and moshi for the actual `LobstersPost` item.

Do we? With your revert I crash with this exception

```
kotlin.reflect.jvm.internal.KotlinReflectionInternalError: This callable does not support a default call: public constructor Submitter(username: kotlin.String, @com.squareup.moshi.Json createdAt: kotlin.String, @com.squareup.moshi.Json isAdmin: kotlin.Boolean, about: kotlin.String, @com.squareup.moshi.Json isModerator: kotlin.Boolean, karma: kotlin.Long = ..., @com.squareup.moshi.Json avatarUrl: kotlin.String, @com.squareup.moshi.Json invitedByUser: kotlin.String, @com.squareup.moshi.Json githubUsername: kotlin.String? = ..., @com.squareup.moshi.Json twitterUsername: kotlin.String? = ..., @com.squareup.moshi.Json keybaseSignatures: kotlin.collections.List<dev.msfjarvis.lobsters.model.KeybaseSignature> = ...) defined in dev.msfjarvis.lobsters.model.Submitter[DeserializedClassConstructorDescriptor@42b19c2]
```

Switching back to codegen fixes it

File: app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt (hunk 2 above)
Comment (Skrilltrax) [reply app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt:18]: We are fetching the list directly in `PagingSource` and there's no underlying list source that we can update and then call `invalidate` on. If we don't go the DB route, we'll need something to back the list and the changes to that backing property will then invalidate the `PagingSource` and update the UI.

File: database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt (hunk 5 above)
Comment (Skrilltrax) [reply database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt:7]: Oh ohkay, I just took a quick look on the generated model.

File: app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt (hunk 2 above)
Comment (msfjarvis) [reply app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt:18]: Paging already handles updating the UI though? That's its entire point.

**Comment by Skrilltrax** [comment]
That's weird, with *only* codegen, The app was throwing an exception something along the lines of 
`Cannot convert List<LobstersPost>: serialising or deserialising without KotlinJsonAdapterFactory() will produce unstable results`

This seemed like a codegen problem since we cannot use codegen on the generated model.

**Comment by msfjarvis** [comment]
> That's weird, with _only_ codegen, The app was throwing an exception something along the lines of
> `Cannot convert List<LobstersPost>: serialising or deserialising without KotlinJsonAdapterFactory() will produce unstable results`
> 
> This seemed like a codegen problem since we cannot use codegen on the generated model.

Yeah I'm confused as well as to why codegen works for me and reflection doesn't.

File: app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt (hunk 2 above)
Comment (Skrilltrax) [reply app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt:18]: AFAIK there's two cases in which Paging would update the UI
 
- When we reach the end of the page
- When there's a change in underlying data source

In the second case, we don't really have a source to update once we the `is_saved` property is updated and hence the `PagedList` never actually changes.

Comment (msfjarvis) [reply app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt:18]: Hm yeah going through the docs it does seem to require a concrete local source as well. Let's add a database for posts but clear it on ViewModel creation so it's empty when paging starts loading.

**Comment by Skrilltrax** [comment]
I'll complete it over the weekend :confused:

**Comment by msfjarvis** [comment]
> I'll complete it over the weekend

:+1:
I've rebased over develop.

File: app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt [hunk 6]
```diff
@@ -17,17 +19,21 @@ import javax.inject.Inject
...
+  private val lobstersApi: LobstersApi,
 ) : ViewModel() {
   private val _savedPosts = MutableStateFlow<List<LobstersPost>>(emptyList())
+  private var source: LobstersPagingSource? = null
   val savedPosts = _savedPosts.asStateFlow()
-  val posts = Pager(PagingConfig(5)) {
-    pagingSource
-  }.flow
+  val posts = Pager(PagingConfig(25)) {
+    source = LobstersPagingSource(lobstersApi, lobstersRepository)
```
Comment (msfjarvis) [inline app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt:28]: why not initialize this directly?

**Comment by Skrilltrax** [comment]
This now handles and saves the saved state correctly. However, a complete list refresh takes place every time the favourite icon is pressed. The user experience is quite bad at this point but I think we can improve it with some minor hacks.

AFAIK the Paging library have it's own RV adapter which performs diffing but nothing similar is available for compose atm.

File: app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt (hunk 6 above)
Comment (Skrilltrax) [reply app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt:28]: You mean via DI, like it was being done previously?

Comment (msfjarvis) [reply app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt:28]: yeah either way works, I just don't see the point of having it be a nullable field when all dependencies are available at class init.

Comment (Skrilltrax) [reply app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt:28]: :+1:

File: app/src/main/java/dev/msfjarvis/lobsters/data/repo/LobstersRepository.kt [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+class LobstersRepository @Inject constructor(private val lobstersDatabase: LobstersDatabase) {
+
+  fun isPostSaved(postId: String): Boolean {
+    val post = lobstersDatabase.postQueries.isPostSaved(postId).executeAsOneOrNull()
+    return post != null
+  }
```
Comment (msfjarvis) [inline app/src/main/java/dev/msfjarvis/lobsters/data/repo/LobstersRepository.kt:14]: why are the others suspending but not this?

File: app/src/main/java/dev/msfjarvis/lobsters/ui/main/MainActivity.kt (hunk 3 above)
Comment (msfjarvis) [reply app/src/main/java/dev/msfjarvis/lobsters/ui/main/MainActivity.kt:67]: Still needs fixing.

File: app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt [hunk 8]
```diff
@@ -4,49 +4,62 @@ import androidx.lifecycle.ViewModel
...
+    source
+  }.flow.cachedIn(viewModelScope)
 
-  init {
-    getSavedPosts()
+  fun toggleSave(post: LobstersPost) {
+    viewModelScope.launch {
+      val isSaved = lobstersRepository.isPostSaved(post.short_id)
+      if (isSaved) removeSavedPost(post) else savePost(post)
+      invalidateSource()
```
Comment (msfjarvis) [inline app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt:36]: I'm assuming this invalidation is what is causing the entire list to be redrawn?

File: app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt [hunk 9]
```diff
@@ -4,49 +4,62 @@ import androidx.lifecycle.ViewModel
...
     viewModelScope.launch {
-      savedPostsDao.loadPosts().collectLatest { _savedPosts.value = it }
+      lobstersRepository.getSavedPosts().forEach {
+        _savedPosts.value = _savedPosts.value + it
+      }
     }
   }
 
-  fun savePost(post: LobstersPost) {
+  fun invalidateSource() {
```
Comment (msfjarvis) [inline app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt:48]:
```suggestion
  private fun invalidateSource() {
```
?

File: buildSrc/src/main/java/Dependencies.kt [hunk 10]
```diff
@@ -66,8 +58,9 @@ object Dependencies {
     object Moshi {
 
       private const val version = "1.11.0"
-      const val codegen = "com.squareup.moshi:moshi-kotlin-codegen:$version"
+      const val kotlinReflect = "com.squareup.moshi:moshi-kotlin:$version"
```
Comment (msfjarvis) [inline buildSrc/src/main/java/Dependencies.kt:61]: remove this?

File: database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,20 @@
+package dev.msfjarvis.lobsters.model
+
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.adapter
+import com.squareup.sqldelight.ColumnAdapter
+import dev.zacsweers.moshix.reflect.MetadataKotlinJsonAdapterFactory
+
+@OptIn(ExperimentalStdlibApi::class)
+class SubmitterAdapter : ColumnAdapter<Submitter, String> {
+  private val moshi = Moshi.Builder().add(MetadataKotlinJsonAdapterFactory()).build()
```
Comment (msfjarvis) [inline database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt:10]: Should probably have these adapters be provided by Dagger so we can make the Moshi initialization centralised as well

File: app/src/main/java/dev/msfjarvis/lobsters/data/repo/LobstersRepository.kt (hunk 7 above)
Comment (Skrilltrax) [reply app/src/main/java/dev/msfjarvis/lobsters/data/repo/LobstersRepository.kt]: Fixed

File: app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt (hunk 8 above)
Comment (Skrilltrax) [reply app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt:35]: Yep, I have a few ideas in mind to fix this :thinking:

File: app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt (hunk 9 above)
Comment (Skrilltrax) [reply app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt]: Dun

File: database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt (hunk 11 above)
Comment (Skrilltrax) [reply database/src/main/java/dev/msfjarvis/lobsters/model/SubmitterAdapter.kt:10]: Will do it later after fixing the invalidation issue

File: buildSrc/src/main/java/Dependencies.kt (hunk 10 above)
Comment (Skrilltrax) [reply buildSrc/src/main/java/Dependencies.kt:61]: :+1:

File: app/src/main/java/dev/msfjarvis/lobsters/data/repo/LobstersRepository.kt [hunk 12]
```diff
@@ -2,35 +2,56 @@ package dev.msfjarvis.lobsters.data.repo
...
+  private val savedPostsCache: MutableMap<String, LobstersPost> = mutableMapOf()
+  private val coroutineScope = CoroutineScope(Job() + Dispatchers.IO)
+
+  init {
+    coroutineScope.launch {
+      getAllPosts().forEach {
+        savedPostsCache.putIfAbsent(it.short_id, it)
+      }
+    }
+  }
```
Comment (Skrilltrax) [inline app/src/main/java/dev/msfjarvis/lobsters/data/repo/LobstersRepository.kt:23]: @msfjarvis currently i am creating cached map on class init. However we can also do it in the first DB access. What are your thoughts on it?

Comment (msfjarvis) [reply app/src/main/java/dev/msfjarvis/lobsters/data/repo/LobstersRepository.kt:23]: Caching on init is fine, we want it to be immediately available

**Comment by Skrilltrax** [comment]
I think all that is left at this point is cleaning up the sq file and rewriting tests.

**Review by msfjarvis — APPROVED** [review]
Unhappy about the ceremony around the saved field value but I guess it's unavoidable.

bors r+

**Comment by Skrilltrax** [comment]
Yeah, I guess we can fix it once there is diff support for compose.
